src/provider/form-factory-provider.tsx

```diff
@@ -0,0 +1,138 @@
+import React, { createContext, useContext, useEffect, useRef } from 'react';
+import { type FormField, type FormSchema, type SessionMode } from '../types';
+import { EncounterFormProcessor } from '../processors/encounter/encounter-form-processor';
+import { type LayoutType, useLayoutType, type OpenmrsResource } from '@openmrs/esm-framework';
+import { type FormProcessorConstructor } from '../processors/form-processor';
+import { type FormContextProps } from './form-provider';
+import { validateForm } from './form-factory-helper';
+
+interface FormFactoryProviderContextProps {
+  patient: fhir.Patient;
+  sessionMode: SessionMode;
+  sessionDate: Date;
+  formJson: FormSchema;
+  formProcessors: Record<string, FormProcessorConstructor>;
+  layoutType: LayoutType;
+  workspaceLayout: 'minimized' | 'maximized';
+  visit: OpenmrsResource;
+  location: OpenmrsResource;
+  provider: OpenmrsResource;
+  registerForm: (formId: string, context: FormContextProps, isSubForm: boolean) => void;
+  getSubForms: () => FormContextProps[];
+  getRootForm: () => FormContextProps;
+  setCurrentPage: (page: string) => void;
+  handleConfirmQuestionDeletion?: (question: Readonly<FormField>) => Promise<void>;
+}
+
+interface FormFactoryProviderProps {
+  patient: fhir.Patient;
+  sessionMode: SessionMode;
+  sessionDate: Date;
+  formJson: FormSchema;
+  workspaceLayout: 'minimized' | 'maximized';
+  location: OpenmrsResource;
+  provider: OpenmrsResource;
+  visit: OpenmrsResource;
+  children: React.ReactNode;
+  formSubmissionProps: {
+    isSubmitting: boolean;
+    setIsSubmitting: (isSubmitting: boolean) => void;
+    onSubmit: (data: any) => void;
+    onError: (error: any) => void;
+    handleClose: () => void;
+  };
+  setCurrentPage: (page: string) => void;
+  handleConfirmQuestionDeletion?: (question: Readonly<FormField>) => Promise<void>;
+}
+
+const FormFactoryProviderContext = createContext<FormFactoryProviderContextProps | undefined>(undefined);
+
+export const FormFactoryProvider: React.FC<FormFactoryProviderProps> = ({
+  patient,
+  sessionMode,
+  sessionDate,
+  formJson,
+  workspaceLayout,
+  location,
+  provider,
+  visit,
+  children,
+  formSubmissionProps,
+  setCurrentPage,
+  handleConfirmQuestionDeletion,
+}) => {
+  const rootForm = useRef<FormContextProps>();
+  const subForms = useRef<Record<string, FormContextProps>>({});
```

Comment:
Why are we using Refs here?

Reply:
The reason I used refs is that they won’t trigger a re-render on subsequent state updates. Do you want me to use the standard `useState`?

Reply:
No, I guess since we're not sending them anywhere it's probably ok.